File: ipfs_accelerate_js/src/utils/web_accelerator.ts
// FI: any;
 * Convert: any;
 * Conversi: any;
 * Th: any;
 * Conversi: any;
 */;



// WebG: any;
export interface Props {initialized: re: any;
  bri: any;
  hardware_detec: any;
  enable_resource_p: any;
  initiali: any;
  ipfs_acceler: any;
  bri: any;
  browser_preferen: any;
  bri: any;
  connection_p: any;
  bri: any;}

/** WebAccelerat: any;

Th: any;
t: any;
f: any;

Key features) {
- Automat: any;
- Browser-specific optimizations (Firefox for ((((audio) { any, Edge for (WebNN) {
- Precision) { an) { an: any;
- Resourc) { an: any;
- IP: any;

impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
impo: any;
// Configu: any;
logging.basicConfig(level = logging.INFO, format) { any) { any: any = '%(asctime: a: any;'
logger: any: any: any = loggi: any;
;
// T: any;
try ${$1} catch(error: any): any {logger.warning("Enhanced WebSock: any;"
  HAS_WEBSOCKET: any: any: any = fa: any;};
try ${$1} catch(error: any): any {logger.warning("IPFS accelerati: any;"
  HAS_IPFS: any: any: any = fa: any;}
// Consta: any;
DEFAULT_PORT: any: any: any = 8: any;
DEFAULT_HOST: any: any: any: any: any: any = "127.0.0.1";"
;
class $1 extends $2 {
  /** Mod: any;
  TEXT) {any = "text";"
  VISION) { any: any: any: any: any: any = "vision";"
  AUDIO: any: any: any: any: any: any = "audio";"
  MULTIMODAL: any: any: any: any: any: any = "multimodal";};"
class $1 extends $2 {/** Unifi: any;
  hardwa: any;
  
  function this( this: any:  any: any): any {  any: any): any {: any { any, $1): any { boolean: any: any: any = tr: any;
        $1: number: any: any = 4, $1: Record<$2, $3> = nu: any;
        $1: string: any: any = "chrome", $1: string: any: any: any: any: any: any = "webgpu",;"
        $1: boolean: any: any = true, $1: number: any: any: any = DEFAULT_PO: any;
        $1: string: any: any = DEFAULT_HOST, $1: boolean: any: any = tr: any;
    /** Initiali: any;
    
    A: any;
      enable_resource_p: any;
      max_connecti: any;
      browser_preferen: any;
      default_brow: any;
      default_platf: any;
      enable_i: any;
      websocket_p: any;
      host) { Ho: any;
      enable_heartbeat) { Wheth: any;
    this.enable_resource_pool = enable_resource_p: any;
    this.max_connections = max_connecti: any;
    this.default_browser = default_brow: any;
    this.default_platform = default_platf: any;
    this.enable_ipfs = enable_i: any;
    this.websocket_port = websocket_p: any;
    this.host = h: any;
    this.enable_heartbeat = enable_heartb: any;
    
    // S: any;
    this.browser_preferences = browser_preferences || ${$1}
    
    // Sta: any;
    this.initialized = fa: any;
    this.loop = n: any;
    this.bridge = n: any;
    this.ipfs_model_cache = {}
    this.active_models = {}
    this.connection_pool = [];
    this._shutting_down = fa: any;
    
    // Statist: any;
    this.stats = ${$1}
    
    // Crea: any;
    try ${$1} catch(error) { any) {) { any {) { any {this.loop = async: any;
      async: any;
    this.hardware_detector = n: any;
    if (((($1) {this.hardware_detector = ipfs_accelerate_impl) { an) { an: any;}
    // Impor) { an: any;
    if (((($1) { ${$1} else {this.ipfs_accelerate = nul) { an) { an: any;};
  async $1($2)) { $3 {/** Initialize WebAccelerator with async setup.}
    $1) { boolean) { tru) { an: any;
    if (((($1) {return true}
    try {
      // Create) { an) { an: any;
      if ((($1) {
        this.bridge = await) { an) { an: any;
          port)) { any { any) { any: any = th: any;
          host) {any = th: any;
          enable_heartbeat: any: any: any = th: any;
        )};
        if (((((($1) { ${$1} else {logger.warning("WebSocket bridge) { an) { an: any;"
      if ((($1) { ${$1}");"
      } else {this.available_hardware = ["cpu"];"
        logger) { an) { an: any;
      if ((($1) { ${$1} catch(error) { any)) { any {logger.error(`$1`)}
      return) { an) { an: any;
  
  $1($2) {/** Initializ) { an: any;
    // I: an: any;
    // F: any;
    this.connection_pool = [];};
  async $1($2) {
    /** Ensu: any;
    if (((((($1) {await this.initialize()}
  function this( this) { any): any { any): any { any): any {  any) { any): any { any, $1)) { any { string, input_data: any) {Any, $1: Record<$2, $3> = nu: any;
    
    A: any;
      model_n: any;
      input_d: any;
      options) { Addition: any;
        - precision) { Precision level (4) { a: any;
        - mixed_precision) { Wheth: any;
        - brow: any;
        - platf: any;
        - optimize_for_au: any;
        - use_i: any;
        
    Retu: any;
      Di: any;
    // R: any;
    retu: any;
  
  async _accelerate_async(this: any, $1: string, input_data: Any, $1: Record<$2, $3> = nu: any;
    /** Asy: any;
    
    A: any;
      model_n: any;
      input_d: any;
      options) { Addition: any;
      
    Returns) {
      Di: any;
    // Ensu: any;
    awa: any;
    
    // Defau: any;
    options) { any) { any: any = options || {}
    
    // Determi: any;
    model_type: any: any = (options["model_type"] !== undefin: any;"
    if ((((((($1) {
      model_type) {any = this._get_model_type(model_name) { any) { an) { an: any;}
    // Ge) { an: any;
    hardware_config: any: any = th: any;
    
    // Overri: any;
    platform) { any) { any = (options["platform"] !== undefined ? options["platform"] : (hardware_config["platform"] !== undefin: any;"
    browser: any: any = (options["browser"] !== undefined ? options["browser"] : (hardware_config["browser"] !== undefin: any;"
    precision: any: any = (options["precision"] !== undefined ? options["precision"] : (hardware_config["precision"] !== undefin: any;"
    mixed_precision: any: any = (options["mixed_precision"] !== undefined ? options["mixed_precision"] : (hardware_config["mixed_precision"] !== undefin: any;"
    
    // Firef: any;
    optimize_for_audio: any: any = (options["optimize_for_audio"] !== undefin: any;"
    if (((((($1) {
      optimize_for_audio) {any = tru) { an) { an: any;}
    // Us) { an: any;
    use_ipfs) { any) { any = this.enable_ipfs && (options["use_ipfs"] !== undefin: any;"
    
    // Prepa: any;
    accel_config: any: any: any = ${$1}
    
    // I: an: any;
    if (((((($1) {
      result) {any = this.ipfs_accelerate(model_name) { any) { an) { an: any;}
      // Updat) { an: any;
      this.stats["total_inferences"] += 1;"
      if (((((($1) { ${$1} else {this.stats["fallback_inferences"] += 1;"
        this.stats["errors"] += 1}"
      if ($1) { ${$1} else {this.stats["ipfs_cache_misses"] += 1) { an) { an: any;"
    
    // I) { an: any;
    // Th: any;
    return await this._accelerate_with_bridge(model_name) { a: any;
  
  async _accelerate_with_bridge(this: any, $1)) { any { stri: any;
    /** Accelera: any;
    
    A: any;
      model_n: any;
      input_d: any;
      config) { Accelerati: any;
      
    Returns) {
      Di: any;
    if ((((((($1) {
      logger) { an) { an: any;
      return ${$1}
    // Wai) { an: any;
    connected) { any) { any) { any: any: any: any = await this.bridge.wait_for_connection() {;
    if (((((($1) {
      logger) { an) { an: any;
      return ${$1}
    // Initializ) { an: any;
    platform) { any) { any: any = (config["platform"] !== undefined ? config["platform"] ) { th: any;"
    model_type: any: any = (config["model_type"] !== undefin: any;"
    
    // Prepa: any;
    model_options: any: any: any = ${$1}
    
    // Initiali: any;
    logg: any;
    init_result: any: any = awa: any;
    ;
    if (((((($1) {
      error_msg) { any) { any) { any = (init_result["error"] !== undefined ? init_result["error"] ) { "Unknown error") if ((((init_result else { "No response) { an) { an: any;"
      logger.error(`$1`) {
      this.stats["errors"] += 1;"
      return ${$1}
    // Ru) { an: any;
    logg: any;
    inference_result) { any) { any = awa: any;
    
    // Upda: any;
    this.stats["total_inferences"] += 1;"
    if (((((($1) { ${$1} else {
      error_msg) { any) { any) { any = (inference_result["error"] !== undefined ? inference_result["error"] ) { "Unknown error") if ((((inference_result else {"No response) { an) { an: any;"
      logger.error(`$1`) {
      this.stats["fallback_inferences"] += 1;"
      this.stats["errors"] += 1) { a: any;"
  ;
  function this( this: any:  any: any): any {  any: any): any { any, $1): any { string, $1) { string: any: any = nu: any;
    /** G: any;
    ;
    Args) {
      model_name) { Na: any;
      model_type) { Type of model (optional: any, will be inferred if ((((((!provided) {
      
    Returns) {
      Dict) { an) { an: any;
    // Determin) { an: any;
    if (((($1) {
      model_type) {any = this._get_model_type(model_name) { any) { an) { an: any;}
    // Tr) { an: any;
    if (((($1) {
      try {
        hardware) {any = this.hardware_detector.get_optimal_hardware(model_name) { any) { an) { an: any;
        logge) { an: any;
        if (((((($1) { ${$1} else {
          platform) {any = this) { an) { an: any;}
        // Ge) { an: any;
        browser) { any: any = th: any;
        
    };
        return ${$1} catch(error: any): any {logger.error(`$1`)}
    // Fallba: any;
    platform: any: any: any = th: any;
    browser: any: any = th: any;
    ;
    return ${$1}
  
  $1($2)) { $3 {/** Get optimal browser for ((((((a model type && platform.}
    Args) {
      model_type) { Type) { an) { an: any;
      platform) { Platfor) { an: any;
      
    Retu: any;
      Brows: any;
    // U: any;
    if (((($1) {return this) { an) { an: any;
    if ((($1) {return "edge"  // Edge) { an) { an: any;"
    if ((($1) {
      return) { an) { an: any;
    else if (((($1) {return "chrome"  // Chrome) { an) { an: any;"
    }
    retur) { an: any;
  
  $1($2)) { $3 {/** Determine model type based on model name.}
    Args) {
      model_name) { Na: any;
      
    Returns) {;
      Mod: any;
    model_name_lower: any: any: any = model_na: any;
    
    // Aud: any;
    if ((((((($1) {return ModelType) { an) { an: any;
    if ((($1) {return ModelType) { an) { an: any;
    if ((($1) {return ModelType) { an) { an: any;
    retur) { an: any;
  
  async $1($2) {/** Cle: any;
    this._shutting_down = t: any;}
    // Clo: any;
    if (((($1) {
      try ${$1} catch(error) { any)) { any {logger.error(`$1`)}
    // Clean) { an) { an: any;
    }
    if ((((($1) {
      try ${$1} catch(error) { any)) { any {logger.error(`$1`)}
    this.initialized = fals) { an) { an: any;
    }
    logge) { an: any;
  ;
  function this(this:  any:  any: any:  any: any): any -> Dict[str, Any]) {
    /** G: any;
    
    Retu: any;
      Di: any;
    // A: any;
    if (((($1) {
      bridge_stats) { any) { any) { any) { any = thi) { an: any;
      combined_stats: any: any: any = ${$1}
      retu: any;
    
    }
    retu: any;

// Help: any;
async create_web_accelerator($1): any { Record<$2, $3> = nu: any;
  /** Crea: any;
  
  A: any;
    opti: any;
    
  Returns) {
    Initializ: any;
  options) { any) { any) { any) { any) { any: any = options || {}
  
  accelerator: any: any: any = WebAccelerat: any;
    enable_resource_pool: any: any = (options["enable_resource_pool"] !== undefin: any;"
    max_connections: any: any = (options["max_connections"] !== undefin: any;"
    browser_preferences: any: any = (options["browser_preferences"] !== undefin: any;"
    default_browser: any: any = (options["default_browser"] !== undefin: any;"
    default_platform: any: any = (options["default_platform"] !== undefin: any;"
    enable_ipfs: any: any = (options["enable_ipfs"] !== undefin: any;"
    websocket_port: any: any = (options["websocket_port"] !== undefin: any;"
    host: any: any = (options["host"] !== undefin: any;"
    enable_heartbeat: any: any = (options["enable_heartbeat"] !== undefin: any;"
  );
  
  // Initiali: any;
  success: any: any: any = awa: any;
  if (((((($1) {logger.error("Failed to) { an) { an: any;"
    retur) { an: any;

// Te: any;
async $1($2) {
  /** Te: any;
  // Crea: any;
  accelerator) { any) { any) { any = awa: any;
  if (((((($1) {logger.error("Failed to) { an) { an: any;"
    return false}
  try {logger.info("WebAccelerator create) { an: any;"
    logg: any;
    text_result) { any) { any: any = accelerat: any;
      "bert-base-uncased",;"
      "This i: an: any;"
      options) { any: any: any: any: any: any = ${$1}
    );
    
}
    logg: any;
    
    // Te: any;
    logg: any;
    audio_result: any: any: any = accelerat: any;
      "openai/whisper-tiny",;"
      ${$1},;
      options: any: any: any: any: any: any = ${$1}
    );
    
    logg: any;
    
    // G: any;
    stats: any: any: any = accelerat: any;
    logg: any;
    
    // Shutd: any;
    awa: any;
    logg: any;
    retu: any;
    ;
  } catch(error: any): any {logger.error(`$1`);
    awa: any;
    return false}
if (((((($1) {
  // Run) { an) { an: any;
  impor) { an: any;
  success) { any) { any = asyn: any;
  s: an: any;

File: tests/test/fixed_web_platform/web_accelerator.py
#!/usr/bin/env python3
"""
WebAccelerator - Unified WebNN/WebGPU Hardware Acceleration

This module provides a unified WebAccelerator class for browser-based WebNN and WebGPU 
hardware acceleration with IPFS content delivery integration. It automatically selects
the optimal browser and hardware backend based on model type and provides a simple API
for hardware-accelerated inference.

Key features:
- Automatic hardware selection based on model type
- Browser-specific optimizations (Firefox for audio, Edge for WebNN)
- Precision control (4-bit, 8-bit, 16-bit) with mixed precision
- Resource pooling for efficient connection reuse
- IPFS integration for model loading
"""

import os
import sys
import json
import time
import asyncio
import logging
import tempfile
import platform as platform_module
from pathlib import Path
from typing import Dict, List, Any, Optional, Union, Tuple, Callable

# Configure logging
logging.basicConfig(level=logging.INFO, format='%(asctime)s - %(levelname)s - %(message)s')
logger = logging.getLogger(__name__)

# Try to import required components
try:
    # Import enhanced WebSocket bridge
    from fixed_web_platform.enhanced_websocket_bridge import EnhancedWebSocketBridge, create_enhanced_websocket_bridge
    HAS_WEBSOCKET = True
except ImportError:
    logger.warning("Enhanced WebSocket bridge not available")
    HAS_WEBSOCKET = False

try:
    # Import IPFS module
    import ipfs_accelerate_impl
    HAS_IPFS = True
except ImportError:
    logger.warning("IPFS acceleration module not available")
    HAS_IPFS = False

# Constants
DEFAULT_PORT = 8765
DEFAULT_HOST = "127.0.0.1"

class ModelType:
    """Model type constants for WebAccelerator."""
    TEXT = "text"
    VISION = "vision"
    AUDIO = "audio"
    MULTIMODAL = "multimodal"

class WebAccelerator:
    """
    Unified WebNN/WebGPU hardware acceleration with IPFS integration.
    
    This class provides a high-level interface for browser-based WebNN and WebGPU
    hardware acceleration with automatic hardware selection, browser-specific 
    optimizations, and IPFS content delivery integration.
    """
    
    def __init__(self, enable_resource_pool: bool = True, 
                 max_connections: int = 4, browser_preferences: Dict[str, str] = None,
                 default_browser: str = "chrome", default_platform: str = "webgpu",
                 enable_ipfs: bool = True, websocket_port: int = DEFAULT_PORT,
                 host: str = DEFAULT_HOST, enable_heartbeat: bool = True):
        """
        Initialize WebAccelerator with configuration.
        
        Args:
            enable_resource_pool: Whether to enable connection pooling
            max_connections: Maximum number of concurrent browser connections
            browser_preferences: Dict mapping model types to preferred browsers
            default_browser: Default browser to use
            default_platform: Default platform to use (webnn or webgpu)
            enable_ipfs: Whether to enable IPFS content delivery
            websocket_port: Port for WebSocket server
            host: Host to bind to
            enable_heartbeat: Whether to enable heartbeat for connection health
        """
        self.enable_resource_pool = enable_resource_pool
        self.max_connections = max_connections
        self.default_browser = default_browser
        self.default_platform = default_platform
        self.enable_ipfs = enable_ipfs
        self.websocket_port = websocket_port
        self.host = host
        self.enable_heartbeat = enable_heartbeat
        
        # Set default browser preferences if not provided
        self.browser_preferences = browser_preferences or {
            ModelType.AUDIO: "firefox",       # Firefox for audio models (optimized compute shaders)
            ModelType.VISION: "chrome",       # Chrome for vision models
            ModelType.TEXT: "edge",           # Edge for text models (WebNN support)
            ModelType.MULTIMODAL: "chrome",   # Chrome for multimodal models
        }
        
        # State variables
        self.initialized = False
        self.loop = None
        self.bridge = None
        self.ipfs_model_cache = {}
        self.active_models = {}
        self.connection_pool = []
        self._shutting_down = False
        
        # Statistics
        self.stats = {
            "total_inferences": 0,
            "total_model_loads": 0,
            "accelerated_inferences": 0,
            "fallback_inferences": 0,
            "ipfs_cache_hits": 0,
            "ipfs_cache_misses": 0,
            "browser_connections": 0,
            "errors": 0
        }
        
        # Create event loop for async operations
        try:
            self.loop = asyncio.get_event_loop()
        except RuntimeError:
            self.loop = asyncio.new_event_loop()
            asyncio.set_event_loop(self.loop)
        
        # Initialize hardware detector if IPFS module is available
        self.hardware_detector = None
        if HAS_IPFS and hasattr(ipfs_accelerate_impl, "HardwareDetector"):
            self.hardware_detector = ipfs_accelerate_impl.HardwareDetector()
            
        # Import IPFS acceleration functions if available
        if HAS_IPFS:
            self.ipfs_accelerate = ipfs_accelerate_impl.accelerate
        else:
            self.ipfs_accelerate = None
    
    async def initialize(self) -> bool:
        """
        Initialize WebAccelerator with async setup.
        
        Returns:
            bool: True if initialization succeeded, False otherwise
        """
        if self.initialized:
            return True
            
        try:
            # Create WebSocket bridge
            if HAS_WEBSOCKET:
                self.bridge = await create_enhanced_websocket_bridge(
                    port=self.websocket_port,
                    host=self.host,
                    enable_heartbeat=self.enable_heartbeat
                )
                
                if not self.bridge:
                    logger.error("Failed to create WebSocket bridge")
                    return False
                    
                logger.info(f"WebSocket bridge created on {self.host}:{self.websocket_port}")
            else:
                logger.warning("WebSocket bridge not available, using simulation")
            
            # Detect hardware capabilities
            if self.hardware_detector:
                self.available_hardware = self.hardware_detector.detect_hardware()
                logger.info(f"Detected hardware: {', '.join(self.available_hardware)}")
            else:
                self.available_hardware = ["cpu"]
                logger.warning("Hardware detector not available, using CPU only")
            
            # Initialize connection pool if enabled
            if self.enable_resource_pool:
                self._initialize_connection_pool()
            
            self.initialized = True
            logger.info("WebAccelerator initialized successfully")
            return True
            
        except Exception as e:
            logger.error(f"Error initializing WebAccelerator: {e}")
            return False
    
    def _initialize_connection_pool(self):
        """Initialize connection pool for browser connections."""
        # In a full implementation, this would set up a connection pool
        # For now, just initialize an empty list
        self.connection_pool = []
    
    async def _ensure_initialization(self):
        """Ensure WebAccelerator is initialized."""
        if not self.initialized:
            await self.initialize()
    
    def accelerate(self, model_name: str, input_data: Any, options: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Accelerate inference with optimal hardware selection.
        
        Args:
            model_name: Name of the model
            input_data: Input data for inference
            options: Additional options for acceleration
                - precision: Precision level (4, 8, 16, 32)
                - mixed_precision: Whether to use mixed precision
                - browser: Specific browser to use
                - platform: Specific platform to use (webnn, webgpu)
                - optimize_for_audio: Enable Firefox audio optimizations
                - use_ipfs: Enable IPFS content delivery
                
        Returns:
            Dict with acceleration results
        """
        # Run async accelerate in the event loop
        return self.loop.run_until_complete(self._accelerate_async(model_name, input_data, options))
    
    async def _accelerate_async(self, model_name: str, input_data: Any, options: Dict[str, Any] = None) -> Dict[str, Any]:
        """
        Async implementation of accelerate.
        
        Args:
            model_name: Name of the model
            input_data: Input data for inference
            options: Additional options for acceleration
            
        Returns:
            Dict with acceleration results
        """
        # Ensure initialization
        await self._ensure_initialization()
        
        # Default options
        options = options or {}
        
        # Determine model type based on model name
        model_type = options.get("model_type")
        if not model_type:
            model_type = self._get_model_type(model_name)
        
        # Get optimal hardware configuration
        hardware_config = self.get_optimal_hardware(model_name, model_type)
        
        # Override with options if specified
        platform = options.get("platform", hardware_config.get("platform"))
        browser = options.get("browser", hardware_config.get("browser"))
        precision = options.get("precision", hardware_config.get("precision", 16))
        mixed_precision = options.get("mixed_precision", hardware_config.get("mixed_precision", False))
        
        # Firefox audio optimizations
        optimize_for_audio = options.get("optimize_for_audio", False)
        if model_type == ModelType.AUDIO and browser == "firefox" and not options.get("optimize_for_audio", None):
            optimize_for_audio = True
        
        # Use IPFS if enabled and not disabled in options
        use_ipfs = self.enable_ipfs and options.get("use_ipfs", True)
        
        # Prepare acceleration configuration
        accel_config = {
            "platform": platform,
            "browser": browser,
            "precision": precision,
            "mixed_precision": mixed_precision,
            "use_firefox_optimizations": optimize_for_audio,
            "model_type": model_type
        }
        
        # If using IPFS, accelerate with IPFS
        if use_ipfs and self.ipfs_accelerate:
            result = self.ipfs_accelerate(model_name, input_data, accel_config)
            
            # Update statistics
            self.stats["total_inferences"] += 1
            if result.get("status") == "success":
                self.stats["accelerated_inferences"] += 1
            else:
                self.stats["fallback_inferences"] += 1
                self.stats["errors"] += 1
            
            if result.get("ipfs_cache_hit", False):
                self.stats["ipfs_cache_hits"] += 1
            else:
                self.stats["ipfs_cache_misses"] += 1
                
            return result
        
        # If IPFS not available, use direct WebNN/WebGPU acceleration
        # This is a simplified implementation that uses the WebSocket bridge
        return await self._accelerate_with_bridge(model_name, input_data, accel_config)
    
    async def _accelerate_with_bridge(self, model_name: str, input_data: Any, config: Dict[str, Any]) -> Dict[str, Any]:
        """
        Accelerate with WebSocket bridge.
        
        Args:
            model_name: Name of the model
            input_data: Input data for inference
            config: Acceleration configuration
            
        Returns:
            Dict with acceleration results
        """
        if not self.bridge:
            logger.error("WebSocket bridge not available")
            return {"status": "error", "error": "WebSocket bridge not available"}
        
        # Wait for bridge connection
        connected = await self.bridge.wait_for_connection()
        if not connected:
            logger.error("WebSocket bridge not connected")
            return {"status": "error", "error": "WebSocket bridge not connected"}
        
        # Initialize model
        platform = config.get("platform", self.default_platform)
        model_type = config.get("model_type", self._get_model_type(model_name))
        
        # Prepare model options
        model_options = {
            "precision": config.get("precision", 16),
            "mixed_precision": config.get("mixed_precision", False),
            "optimize_for_audio": config.get("use_firefox_optimizations", False)
        }
        
        # Initialize model in browser
        logger.info(f"Initializing model {model_name} with {platform}")
        init_result = await self.bridge.initialize_model(model_name, model_type, platform, model_options)
        
        if not init_result or init_result.get("status") != "success":
            error_msg = init_result.get("error", "Unknown error") if init_result else "No response"
            logger.error(f"Failed to initialize model {model_name}: {error_msg}")
            self.stats["errors"] += 1
            return {"status": "error", "error": error_msg, "model_name": model_name}
        
        # Run inference
        logger.info(f"Running inference with model {model_name} on {platform}")
        inference_result = await self.bridge.run_inference(model_name, input_data, platform, model_options)
        
        # Update statistics
        self.stats["total_inferences"] += 1
        if inference_result and inference_result.get("status") == "success":
            self.stats["accelerated_inferences"] += 1
        else:
            error_msg = inference_result.get("error", "Unknown error") if inference_result else "No response"
            logger.error(f"Failed to run inference with model {model_name}: {error_msg}")
            self.stats["fallback_inferences"] += 1
            self.stats["errors"] += 1
        
        return inference_result
    
    def get_optimal_hardware(self, model_name: str, model_type: str = None) -> Dict[str, Any]:
        """
        Get optimal hardware for a model.
        
        Args:
            model_name: Name of the model
            model_type: Type of model (optional, will be inferred if not provided)
            
        Returns:
            Dict with optimal hardware configuration
        """
        # Determine model type if not provided
        if not model_type:
            model_type = self._get_model_type(model_name)
        
        # Try to use hardware detector if available
        if self.hardware_detector and hasattr(self.hardware_detector, "get_optimal_hardware"):
            try:
                hardware = self.hardware_detector.get_optimal_hardware(model_name, model_type)
                logger.info(f"Optimal hardware for {model_name} ({model_type}): {hardware}")
                
                # Determine platform based on hardware
                if hardware in ["webgpu", "webnn"]:
                    platform = hardware
                else:
                    platform = self.default_platform
                
                # Get browser based on model type and platform
                browser = self._get_browser_for_model(model_type, platform)
                
                return {
                    "hardware": hardware,
                    "platform": platform,
                    "browser": browser,
                    "precision": 16,  # Default precision
                    "mixed_precision": False  # Default to no mixed precision
                }
            except Exception as e:
                logger.error(f"Error getting optimal hardware: {e}")
        
        # Fallback to default configuration
        platform = self.default_platform
        browser = self._get_browser_for_model(model_type, platform)
        
        return {
            "hardware": platform,
            "platform": platform,
            "browser": browser,
            "precision": 16,
            "mixed_precision": False
        }
    
    def _get_browser_for_model(self, model_type: str, platform: str) -> str:
        """
        Get optimal browser for a model type and platform.
        
        Args:
            model_type: Type of model
            platform: Platform to use
            
        Returns:
            Browser name
        """
        # Use browser preferences if available
        if model_type in self.browser_preferences:
            return self.browser_preferences[model_type]
        
        # Use platform-specific defaults
        if platform == "webnn":
            return "edge"  # Edge has best WebNN support
        
        # For WebGPU, use model-specific optimizations
        if model_type == ModelType.AUDIO:
            return "firefox"  # Firefox has best audio performance
        elif model_type == ModelType.VISION:
            return "chrome"  # Chrome has good vision performance
        
        # Default browser
        return self.default_browser
    
    def _get_model_type(self, model_name: str) -> str:
        """
        Determine model type based on model name.
        
        Args:
            model_name: Name of the model
            
        Returns:
            Model type
        """
        model_name_lower = model_name.lower()
        
        # Audio models
        if any(x in model_name_lower for x in ["whisper", "wav2vec", "clap", "audio"]):
            return ModelType.AUDIO
        
        # Vision models
        if any(x in model_name_lower for x in ["vit", "clip", "detr", "image", "vision"]):
            return ModelType.VISION
        
        # Multimodal models
        if any(x in model_name_lower for x in ["llava", "xclip", "multimodal"]):
            return ModelType.MULTIMODAL
        
        # Default to text
        return ModelType.TEXT
    
    async def shutdown(self):
        """Clean up resources and shutdown."""
        self._shutting_down = True
        
        # Close WebSocket bridge
        if self.bridge:
            try:
                # Send shutdown command to browser
                await self.bridge.shutdown_browser()
                
                # Stop WebSocket server
                await self.bridge.stop()
                logger.info("WebSocket bridge stopped")
            except Exception as e:
                logger.error(f"Error stopping WebSocket bridge: {e}")
        
        # Clean up connection pool
        if self.enable_resource_pool and self.connection_pool:
            try:
                for connection in self.connection_pool:
                    # In a full implementation, this would close each connection
                    pass
                logger.info("Connection pool cleaned up")
            except Exception as e:
                logger.error(f"Error cleaning up connection pool: {e}")
        
        self.initialized = False
        logger.info("WebAccelerator shutdown complete")
    
    def get_stats(self) -> Dict[str, Any]:
        """
        Get acceleration statistics.
        
        Returns:
            Dict with acceleration statistics
        """
        # Add bridge stats if available
        if self.bridge:
            bridge_stats = self.bridge.get_stats()
            combined_stats = {
                **self.stats,
                "bridge": bridge_stats
            }
            return combined_stats
        
        return self.stats

# Helper function to create and initialize WebAccelerator
async def create_web_accelerator(options: Dict[str, Any] = None) -> Optional[WebAccelerator]:
    """
    Create and initialize a WebAccelerator instance.
    
    Args:
        options: Configuration options for WebAccelerator
        
    Returns:
        Initialized WebAccelerator or None if initialization failed
    """
    options = options or {}
    
    accelerator = WebAccelerator(
        enable_resource_pool=options.get("enable_resource_pool", True),
        max_connections=options.get("max_connections", 4),
        browser_preferences=options.get("browser_preferences"),
        default_browser=options.get("default_browser", "chrome"),
        default_platform=options.get("default_platform", "webgpu"),
        enable_ipfs=options.get("enable_ipfs", True),
        websocket_port=options.get("websocket_port", DEFAULT_PORT),
        host=options.get("host", DEFAULT_HOST),
        enable_heartbeat=options.get("enable_heartbeat", True)
    )
    
    # Initialize accelerator
    success = await accelerator.initialize()
    if not success:
        logger.error("Failed to initialize WebAccelerator")
        return None
    
    return accelerator

# Test function for WebAccelerator
async def test_web_accelerator():
    """Test WebAccelerator functionality."""
    # Create and initialize WebAccelerator
    accelerator = await create_web_accelerator()
    if not accelerator:
        logger.error("Failed to create WebAccelerator")
        return False
    
    try:
        logger.info("WebAccelerator created successfully")
        
        # Test with a text model
        logger.info("Testing with text model...")
        text_result = accelerator.accelerate(
            "bert-base-uncased",
            "This is a test",
            options={
                "precision": 8,
                "mixed_precision": True
            }
        )
        
        logger.info(f"Text model result: {json.dumps(text_result, indent=2)}")
        
        # Test with an audio model
        logger.info("Testing with audio model...")
        audio_result = accelerator.accelerate(
            "openai/whisper-tiny",
            {"audio": "test.mp3"},
            options={
                "browser": "firefox",
                "optimize_for_audio": True
            }
        )
        
        logger.info(f"Audio model result: {json.dumps(audio_result, indent=2)}")
        
        # Get statistics
        stats = accelerator.get_stats()
        logger.info(f"WebAccelerator stats: {json.dumps(stats, indent=2)}")
        
        # Shutdown
        await accelerator.shutdown()
        logger.info("WebAccelerator test completed successfully")
        return True
        
    except Exception as e:
        logger.error(f"Error in WebAccelerator test: {e}")
        await accelerator.shutdown()
        return False

if __name__ == "__main__":
    # Run test if script executed directly
    import asyncio
    success = asyncio.run(test_web_accelerator())
    sys.exit(0 if success else 1)